### PR TITLE
Enable setting VNet peering properties

### LINKED ISSUE #3183: VPN gateway transit for virtual network peering
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

In AzureCluster resource we can add VNet peering from cluster's VNet to any other VNet. The current implementation creates peering with default properties.

In order to [configure VPN gateway transit for virtual network peering](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit), [additional non-default VNet peering properties have to be set](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit#to-add-a-peering-and-enable-transit).

There are two gateway transit scenarios from the point of view of a CAPZ cluster, depending on where the VPN gateway is deployed:
1. VPN gateway is deployed to the remote peered VNet
2. VPN gateway is deployed to the CAPZ cluster VNet (this feature does not exist in CAPZ, but it can be done outside of CAPZ)

In both scenarios, the following properties are set on VNet peering on both CAPZ and peered VNets:
- Traffic to remote virtual network: Allow
- Traffic forwarded from remote virtual network: Allow

Furthermore, depending on where the VPN gateway is deployed, peering on both sides has to be configured with correct Virtual network gateway option.

1\. VPN gateway is deployed to the remote peered VNet

The following property is set on VNet peering in remote peered VNet:
- Virtual network gateway: Use this virtual network's gateway

The following property is set on VNet peering in CAPZ cluster VNet:
- Virtual network gateway: Use the remote virtual network's gateway

2\. VPN gateway is deployed to the CAPZ cluster VNet

The following property is set on VNet peering in CAPZ cluster VNet:
- Virtual network gateway: Use this virtual network's gateway

The following property is set on VNet peering in remote peered VNet:
- Virtual network gateway: Use the remote virtual network's gateway

In order to implement this, the following VNet peering properties (from Azure SDK) should be exposed in `VNetPeeringSpec` (currently CAPZ is not setting these):
- `AllowForwardedTraffic`
- `AllowGatewayTransit`
- `UseRemoteGateways`

I'm not sure if `AllowVirtualNetworkAccess` is required in order to implement hub and spoke architecture described [in Azure docs here](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit), so that one might be required as well.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

With this implemented, it becomes possible to implement [gateway transit with hub and spoke architecture](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit), where e.g. management cluster VNet is the hub, and the workload cluster VNets are the spokes.

Deploying VPN gateway is not in the scope of this issue, as it can also be deployed independently of CAPZ and it does not have to be deployed in any Kubernetes cluster VNet at all. That being said, CAPZ deploying VPN gateway would be very useful as it would enable implementation of this and many other architectures, so I will open another issue for that and link it here.

### PR DESCRIPTION
**Pull request by nprokopic**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR enables setting additional properties for VNet peering (`AllowForwardedTraffic`, `AllowGatewayTransit`, `AllowVirtualNetworkAccess` and `UseRemoteGateways`).

One use case for setting these properties to non-default values is to, for example, [configure VPN gateway transit for virtual network peering](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3183

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Enable setting VNet peering properties (`AllowForwardedTraffic`, `AllowGatewayTransit`, `AllowVirtualNetworkAccess` and `UseRemoteGateways`)
```

### CODE REVIEW

**Review by willie-yao — COMMENTED** [review]
This looks great, thanks for working on this @nprokopic !

Just one comment from my end: would you be able to add a unit test for `VnetPeeringSpecs()` in `cluster_test.go`?

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
...
+	// +optional
+	AllowGatewayTransit *bool `json:"allowGatewayTransit,omitempty"`
+
+	// AllowVirtualNetworkAccess specifies whether the VMs in the local virtual network space would be able to access
+	// the VMs in remote virtual network space.
+	// +optional
+	AllowVirtualNetworkAccess *bool `json:"allowVirtualNetworkAccess,omitempty"`
+
+	// UseRemoteGateways specifies if remote gateways can be used on this virtual network.
+	//
```
Comment (Jont828) [inline api/v1beta1/types.go:157]: Can probably remove this extra line

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
...
+	// If the flag is set to true, and allowGatewayTransit on remote peering is also true, virtual network will use
```
Comment (Jont828) [inline api/v1beta1/types.go:158]:
```suggestion
	// If the flag is set to true, and allowGatewayTransit on remote peering is also set to true, the virtual network will use
```

File: api/v1beta1/types.go [hunk 3]
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
...
+	// gateways of remote virtual network for transit. Only one peering can have this flag set to true. This flag cannot
```
Comment (Jont828) [inline api/v1beta1/types.go:159]:
```suggestion
	// the gateways of the remote virtual network for transit. Only one peering can have this flag set to true. This flag cannot
```

File: azure/services/vnetpeerings/vnetpeerings_test.go [hunk 4]
```diff
@@ -65,6 +66,30 @@ var (
...
 		SubscriptionID:      "sub1",
 	}
+	fakePeeringHubToSpoke = VnetPeeringSpec{
+		PeeringName:               "hub-to-spoke",
+		SourceVnetName:            "hub-vnet",
+		SourceResourceGroup:       "hub-group",
+		RemoteVnetName:            "spoke-vnet",
+		RemoteResourceGroup:       "spoke-group",
+		SubscriptionID:            "sub1",
+		AllowForwardedTraffic:     to.BoolPtr(true),
```
Comment (Jont828) [inline azure/services/vnetpeerings/vnetpeerings_test.go:76]:
```suggestion
		AllowForwardedTraffic:    pointer.Bool(true),
```
I think we want to switch to the `"k8s.io/utils/pointer"` library

**Review by Jont828 — COMMENTED** [review]
Thanks for the contribution! As @willie-yao said, we might want to add some tests in cluster_tests.go, and here are a couple points from me.

**Comment by nprokopic** [comment]
@Jont828 @willie-yao I have added missing tests and also addressed other suggestions from the review. All tests are :green_circle: :)

Can we add this PR to the `v1.9` milestone?

**Comment by Jont828** [comment]
/lgtm

This looks great! Unless there are any objections, I don't see why not.

**Comment by nawazkh** [comment]
Sweet! Thanks for adding this! 🚀

**Comment by nprokopic** [comment]
> @willie-yao: You must be a member of the [kubernetes-sigs/cluster-api-provider-azure-maintainers](https://github.com/orgs/kubernetes-sigs/teams/cluster-api-provider-azure-maintainers/members) GitHub team to set the milestone. If you believe you should be able to issue the /milestone command, please contact your Cluster API Provider Azure Maintainers and have them propose you as an additional delegate for this responsibility.

Can somebody from @kubernetes-sigs/cluster-api-provider-azure-maintainers help out here?

**Comment by CecileRobertMichon** [comment]
/approve

Thanks all for the reviews 🚀

**Pull request merged**